lib/elixir/lib/calendar/iso.ex

```diff
@@ -663,6 +664,166 @@ defmodule Calendar.ISO do
     end
   end
 
+  @doc """
+  Parses an ISO 8601 formatted duration string to a list of `Duration` compabitble unit pairs.
+
+  See `Duration.from_iso8601/1`.
+
+  """
+  @doc since: "1.17.0"
+  @spec parse_duration(String.t()) :: {:ok, [Duration.unit_pair()]} | {:error, atom()}
+  def parse_duration(string) do
```

Comment:
Here is another implementation, which should be faster because we build less intermediate tuples (but I didn't benchmark it):

```elixir
  @spec parse_duration(String.t()) :: {:ok, [Duration.unit_pair()]} | {:error, atom()}
  def parse_duration("P" <> string) when byte_size(string) > 0 do
    parse_duration_date(string, [], year: ?Y, month: ?M, week: ?W, day: ?D)
  end

  def parse_duration(_) do
    {:error, :invalid_duration}
  end

  defp parse_duration_date("", acc, _allowed), do: {:ok, acc}

  defp parse_duration_date("T" <> string, acc, _allowed) when byte_size(string) > 0 do
    parse_duration_time(string, acc, hour: ?H, minute: ?M, second: ?S)
  end

  defp parse_duration_date(string, acc, allowed) do
    with {integer, <<next, rest::binary>>} <- Integer.parse(string),
         {key, allowed} <- find_unit(allowed, next) do
      parse_duration_date(rest, [{key, integer} | acc], allowed)
    else
      _ -> {:error, :invalid_date_component}
    end
  end

  defp parse_duration_time("", acc, _allowed), do: {:ok, acc}

  defp parse_duration_time(string, acc, allowed) do
    case Integer.parse(string) do
      {second, <<?., rest::binary>>} ->
        case parse_microsecond(rest) do
          {{ms, precision}, "S"} ->
            {:ok, [second: second, microsecond: {ms, precision}] ++ acc}

          _ ->
            {:error, :invalid_time_component}
        end

      {integer, <<next, rest::binary>>} ->
        case find_unit(allowed, next) do
          {key, allowed} -> parse_duration_time(rest, [{key, integer} | acc], allowed)
          false -> {:error, :invalid_time_component}
        end

      _ ->
        {:error, :invalid_time_component}
    end
  end

  defp find_unit([{key, unit} | rest], unit), do: {key, rest}
  defp find_unit([_ | rest], unit), do: find_unit(rest, unit)
  defp find_unit([], _unit), do: false
```

Two differences are:

1. We don't support `,` (because we don't support it anywhere else)
2. We require the decimal part to be present

Reply:
It looks like we do explicitly support a comma though https://github.com/elixir-lang/elixir/blob/main/lib/elixir/lib/calendar/iso.ex#L1725 and the spec also explicitly allows it, so i would lean towards supporting it in duration strings if we can

Reply:
Ah, my bad. So let's support the comma. It should be a small change to the snippet above. :)

Reply:
as an aside, your version outperforms all the previous iterations :)

```
Name                    ips        average  deviation         median         99th %
Elixir parser      368.93 K        2.71 μs  ±1671.85%        2.21 μs        4.17 μs
Regex               38.56 K       25.94 μs    ±83.80%       23.88 μs       39.88 μs

Comparison:
Elixir parser      368.93 K
Regex               38.56 K - 9.57x slower +23.23 μs

Memory usage statistics:

Name             Memory usage
Elixir parser         8.19 KB
Regex                20.62 KB - 2.52x memory usage +12.43 KB
```